3.3.5 gameobject summoner

jackpoz opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

-  Add support to TempSummon to have GameObject as owner (before it was only Unit). The owner actually becomes of type WorldObject.

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Issues addressed:** Closes #  (insert issue tracker number)
Closes #23242
Closes #23131

**Tests performed:** (Does it build, tested in-game, etc.)
It builds, that's it.

**Known issues and TODO list:** (add/remove lines as needed)
- [ ] All the previous behavior with Unit as summoner should be tested
- [ ] The new behavior with GameObject as summoner should be tested
- [ ] Check no-pch build

<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

jackpoz committed:
Scripts/Misc: Change IsSummonedBy(Unit*) to IsSummonedBy(WorldObject*)

jackpoz committed:
Scripts/Misc: Fix build

jackpoz committed:
Core/TempSummons: Rename GetSummoner() to GetSummonerUnit()

jackpoz committed:
Core/TempSummons: Add support to TempSummons::GetSummoner() to return GameObject too

jackpoz committed:
Fix build

jackpoz committed:
Core/TempSummons: Allow GameObject to be owner of TempSummon

illfated commented:
Thank you for your effort & for embarking upon this chapter. Appreciated. 👍

xvwyh commented on src/server/game/Entities/Creature/TemporarySummon.cpp:
```diff
@@ -35,7 +37,12 @@ m_timer(0), m_lifetime(0)
+
+Unit* TempSummon::GetSummonerUnit() const
 {
     return m_summonerGUID ? ObjectAccessor::GetUnit(*this, m_summonerGUID) : nullptr;
```
I'd personally change this to
```cpp
if (WorldObject* summoner = GetSummoner())
    return summoner->ToUnit();
return nullptr;
```
because duplicating logic is never a good idea.

jackpoz replied:
Most of the changes were done in small steps to "migrate" the existing code to the new return types so I tried to change the least possible with each iteration.
It sounds good to do some codestyle changes too, maybe I can polish it up after it's confirmed that it works :)

xvwyh commented on src/server/game/Entities/Creature/TemporarySummon.h:
```diff
@@ -46,7 +46,8 @@ class TC_GAME_API TempSummon : public Creature
         void SaveToDB(uint32 /*mapid*/, uint8 /*spawnMask*/, uint32 /*phaseMask*/) override { }
-        Unit* GetSummoner() const;
+        WorldObject* GetSummoner() const;
+        Unit* GetSummonerUnit() const;
```
Might as well add `GetSummonerGameObject()` for completeness sake and to more clearly communicate that summoner might be one of these two types.

xvwyh commented on src/server/game/Entities/Object/Object.cpp:
```diff
@@ -1884,23 +1884,25 @@ TempSummon* Map::SummonCreature(uint32 entry, Position const& pos, SummonPropert
     if (summoner)
         phase = summoner->GetPhaseMask();
 
+    Unit* summonerUnit = summoner ? summoner->ToUnit() : nullptr;
```
Perhaps merge this with the tiny if branch above?

xvwyh commented on src/server/game/Entities/Object/Object.cpp:
```diff
@@ -1884,23 +1884,25 @@ TempSummon* Map::SummonCreature(uint32 entry, Position const& pos, SummonPropert
             break;
         case UNIT_MASK_GUARDIAN:
-            summon = new Guardian(properties, summoner, false);
+            summon = new Guardian(properties, summonerUnit, false);
```
Such a shame that `Minion` and other derived classes rely on `Unit* Minion::GetOwner()` never being null. And that they have this `Minion::GetOwner()` in the first place, which "conveniently" shadows `Unit::GetOwner()`, no doubt causing a few bugs along the way.

There are at least 2 cases of trap-GOs summoning non-wild creatures in 3.3.5: spells 53076 and 54355. Very curious to see sniffs with what's happening there, but in the context of this PR it's not a big deal, since currently I imagine these spells don't work at all because of a guard condition in the beginning of `Spell::SummonGuardian` that would prevent non-unit spellcasters from summoning any guardians.

jackpoz replied:
It felt like I should not have added GameObject owner support to TempSummon types that I wasn't sure should have (like is a GameObject supposed to be a totem ?) but in the initial local changes I did I actually changed that too.
So if you have a couple of spells that are supposed to spawn a Guardian with a GameObject owner, it's quite easy to add support to that too.

xvwyh replied:
I think it might be easier to force such spells to summon "wild" TempSummons instead of Guardians if the caster isn't a Unit, until it can be proven that they should be Guardians. Client might not react well to it either, what would it even display, "Volatile Trap Knockback Bunny <Volatile Trap's Guardian>"? Would client be able to correctly inherit owner trap's faction for reaction checks? I doubt it. It's not worth the hassle to complicate the code further just for two spells which we aren't even sure how they worked on retail; for all we know, blizzards server could gleefully discard "ALLY/GUARDIAN" summon category if the spell is cast by a GO. And furthermore, they shouldn't currently be working in TC anyway (unless someone bothered to hackfix them) ever since the "GO casting" commit which happened years ago. And if no one raised a stink by now - they won't after this PR either, and it will get fixed when someone will get around to it.

xvwyh replied:
Found a sniff with 53079
<details>

```
ServerToClient: SMSG_UPDATE_OBJECT (0x0C22) Length: 778 ConnIdx: 0 Time: 11/16/2013 13:28:07.000 Number: 1032041
Map: 571
Count: 3
[0] UpdateType: DestroyObjects
[0] Object Count: 1

[2] UpdateType: CreateObject1
[2] GUID: Full: 0xF13072D5004A2348 Type: Creature Entry: 29397 (Land Mine Bunny) Low: 4858696
[2] Object Type: 3 (Unit)
[2] Has Attacking Target: False
[2] Transport Frames Count: 0
[2] Has Stationary Position: False
[2] Is Living: True
[2] Has Anim Kits: False
[2] Has Vehicle Data: False
[2] Has GameObject Position: False
[2] Has Scene Object Data: False
[2] Has GameObject Rotation: False
[2] bitA4: False
[2] Movement Flags: 1536 (OnTransport, DisableGravity)
[2] bit8C: True
[2] bit8D: False
[2] Spline Id: 234583155
[2] FloatBC: 7
[2] FloatAC: 8
[2] FloatA8: 2.5
[2] Time: 1629028945
[2] FloatB0: 4.5
[2] FloatC0: 4.5
[2] FloatC4: 3.141593
[2] FloatC8: 3.141593
[2] FloatB4: 4.72222
[2] FloatB8: 2.5
[2] GUID1: Full: 0xF13072D5004A2348 Type: Creature Entry: 29397 (Land Mine Bunny) Low: 4858696
[2] Position: X: 6116.151 Y: -655.6545 Z: 383.1613
[2] Orientation: 2.076939
[2] OBJECT_FIELD_GUID: 4858696/6.808483E-39
[2] 1: 4046484181/-8.73731E+29
[2] OBJECT_FIELD_TYPE: 9/1.261169E-44
[2] OBJECT_FIELD_ENTRY: 29397/4.119397E-41
[2] OBJECT_FIELD_SCALE_X: 1065353216/1
[2] UNIT_FIELD_BYTES_0: 33554688/9.404242E-38
[2] UNIT_FIELD_DISPLAY_POWER: 1/1.401298E-45
[2] UNIT_FIELD_HEALTH: 11770/1.649328E-41
[2] UNIT_FIELD_MAXHEALTH: 11770/1.649328E-41
[2] UNIT_FIELD_LEVEL: 78/1.093013E-43
[2] UNIT_FIELD_FACTIONTEMPLATE: 14/1.961818E-44
[2] UNIT_FIELD_FLAGS: 33554432/9.403955E-38
[2] UNIT_FIELD_FLAGS_2: 2048/2.869859E-42
[2] UNIT_FIELD_BASEATTACKTIME: 2000/2.802597E-42
[2] UNIT_FIELD_BASEATTACKTIME + 1: 2000/2.802597E-42
[2] UNIT_FIELD_BOUNDINGRADIUS: 1056964608/0.5
[2] UNIT_FIELD_COMBATREACH: 1065353216/1
[2] UNIT_FIELD_DISPLAYID: 11686/1.637557E-41
[2] UNIT_FIELD_NATIVEDISPLAYID: 11686/1.637557E-41
[2] UNIT_FIELD_MOD_CASTING_SPEED: 1065353216/1
[2] UNIT_FIELD_MOD_SPELL_HASTE: 1065353216/1
[2] UNIT_FIELD_MOD_HASTE: 1065353216/1
[2] UNIT_FIELD_MOD_RANGED_HASTE: 1065353216/1
[2] UNIT_FIELD_MOD_HASTE_REGEN: 1065353216/1
[2] UNIT_CREATED_BY_SPELL: 54355/7.616758E-41
[2] UNIT_FIELD_BYTES_2: 1/1.401298E-45
[2] UNIT_FIELD_HOVERHEIGHT: 1065353216/1
```
</details>

The spell is supposed to summon a Guardian, but as we can see here, summoned creatures have no owner GUID among their update fields. They do have the same level and faction as the summoner GO, but we don't have enough information to conclude that they inherit this data from summoner GO, because in CreatureDifficulty this creature 29397 also has level 78 and faction 14 set.

xvwyh replied:
Anyway, my suggestion would be to change Spell::EffectSummonType like this:
```diff
     switch (properties->Category)
     {
         case SUMMON_CATEGORY_WILD:
         case SUMMON_CATEGORY_ALLY:
         case SUMMON_CATEGORY_UNK:
             if (properties->Flags & 512)
             {
                 SummonGuardian(effIndex, entry, properties, numSummons);
                 break;
             }
-            switch (properties->Type)
+            switch (casterUnit ? properties->Type : SUMMON_TYPE_NONE)
             {
                case SUMMON_TYPE_PET:
                case SUMMON_TYPE_GUARDIAN:
                case SUMMON_TYPE_MINION:
                case SUMMON_TYPE_RUNEBLADE:
                case SUMMON_TYPE_CONSTRUCT:
                case SUMMON_TYPE_LIGHTWELL:
                case SUMMON_TYPE_BUTLER:
                    SummonGuardian(effIndex, entry, properties, numSummons);
                    break;
```
and a similar change to the code above the line that this comment is on. This would force GOs to summon wild tempsummons no matter what oddities might be in spell data. Unless Flags&512, that is... but there are no cases of this in 3.3.5, so I dunno.

Killyana commented:
SMART_EVENT_JUST_SUMMONED and SMART_TARGET_OWNER_OR_SUMMONER still not working if the gob is the summoner.

jackpoz commented:
@Killyana yeah I didn't add support to those, see https://github.com/TrinityCore/TrinityCore/pull/23289/files#diff-703f1f16a955ae7ddc7caf673a75d85aR641 . Is that something that you need ?

Killyana commented:
Yes needed in many cases like https://github.com/TrinityCore/TrinityCore/issues/23131

jackpoz committed:
Core/TempSummons: Add support to SAI for GameObject owner of TempSummon

jackpoz commented:
@Killyana I added support to SAI, in particular SMART_TARGET_OWNER_OR_SUMMONER and SMART_EVENT_JUST_SUMMONED.
Let me know if there are other SAI thingies to add (but we can add them later on too)

jackpoz committed:
Scripts/Misc: Fix no-pch build

Killyana commented:
There's also SMART_EVENT_SUMMON_DESPAWNED

jackpoz commented:
SMART_EVENT_SUMMON_DESPAWNED was broken before this PR, fixed it in https://github.com/TrinityCore/TrinityCore/commit/457fc224733cc78cedc5e956442a9f5d7bd2f049 .
I added support for GO owner too already.

jackpoz committed:
ore/TempSummons: Implement PR comments

Killyana commented:
I made this script for a quest, but the event  on Summon despawned doesn't work if the summoner is a gob. I tested with a creature as summoner and it works fine
``` sql
-- Demon Portal SAI
UPDATE `gameobject` SET `spawntimesecs`=120 WHERE `id` IN (177243, 177365, 177366, 177367, 177368, 177369, 177397, 177398, 177399, 177400);
DELETE FROM `smart_scripts` WHERE `entryorguid` IN (177243, 177365, 177366, 177367, 177368, 177369, 177397, 177398, 177399, 177400) AND `source_type`=1;
DELETE FROM `smart_scripts` WHERE `entryorguid` IN (17724300, 17736500, 17736600, 17736700, 17736800, 17736900, 17739700, 17739800, 17739900, 17740000) AND `source_type`=9;
INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`target_o`,`comment`) VALUES
(177243,1,0,0,70,0,100,0,2,0,0,0,80,17724300,2,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on state change - Call Timed ActionList'),
(177243,1,1,0,35,0,100,0,11937,5000,5000,0,106,16,0,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on Summon despawned - Remove flag'),
(177365,1,0,0,70,0,100,0,2,0,0,0,80,17724300,2,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on state change - Call Timed ActionList'),
(177365,1,1,0,35,0,100,0,11937,5000,5000,0,106,16,0,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on Summon despawned - Remove flag'),
(177366,1,0,0,70,0,100,0,2,0,0,0,80,17724300,2,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on state change - Call Timed ActionList'),
(177366,1,1,0,35,0,100,0,11937,5000,5000,0,106,16,0,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on Summon despawned - Remove flag'),
(177367,1,0,0,70,0,100,0,2,0,0,0,80,17724300,2,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on state change - Call Timed ActionList'),
(177367,1,1,0,35,0,100,0,11937,5000,5000,0,106,16,0,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on Summon despawned - Remove flag'),
(177368,1,0,0,70,0,100,0,2,0,0,0,80,17724300,2,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on state change - Call Timed ActionList'),
(177368,1,1,0,35,0,100,0,11937,5000,5000,0,106,16,0,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on Summon despawned - Remove flag'),
(177369,1,0,0,70,0,100,0,2,0,0,0,80,17724300,2,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on state change - Call Timed ActionList'),
(177369,1,1,0,35,0,100,0,11937,5000,5000,0,106,16,0,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on Summon despawned - Remove flag'),
(177397,1,0,0,70,0,100,0,2,0,0,0,80,17724300,2,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on state change - Call Timed ActionList'),
(177397,1,1,0,35,0,100,0,11937,5000,5000,0,106,16,0,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on Summon despawned - Remove flag'),
(177398,1,0,0,70,0,100,0,2,0,0,0,80,17724300,2,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on state change - Call Timed ActionList'),
(177398,1,1,0,35,0,100,0,11937,5000,5000,0,106,16,0,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on Summon despawned - Remove flag'),
(177399,1,0,0,70,0,100,0,2,0,0,0,80,17724300,2,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on state change - Call Timed ActionList'),
(177399,1,1,0,35,0,100,0,11937,5000,5000,0,106,16,0,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on Summon despawned - Remove flag'),
(177400,1,0,0,70,0,100,0,2,0,0,0,80,17724300,2,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on state change - Call Timed ActionList'),
(177400,1,1,0,35,0,100,0,11937,5000,5000,0,106,16,0,0,0,0,0,1,0,0,0,0,0,0,0,'demon portal - on Summon despawned - Remove flag'),
(17724300, 9, 0, 0, 0, 0, 100, 0, 0, 0, 0, 0, 12, 11937, 4, 30000, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,0,'demon portal - ActionList - summon Demon Portal Guardian'),
(17724300, 9, 1, 0, 0, 0, 100, 0, 3500, 3500, 0, 0, 105, 16, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,0,'demon portal - ActionList - Set gob flag');
-- Demon Portal Guardian
DELETE FROM `smart_scripts` WHERE `entryorguid`=11937 AND `source_type`=0;
INSERT INTO `smart_scripts` (`entryorguid`, `source_type`, `id`, `link`, `event_type`, `event_phase_mask`, `event_chance`, `event_flags`, `event_param1`, `event_param2`, `event_param3`, `event_param4`, `action_type`, `action_param1`, `action_param2`, `action_param3`, `action_param4`, `action_param5`, `action_param6`, `target_type`, `target_param1`, `target_param2`, `target_param3`, `target_x`, `target_y`, `target_z`, `target_o`, `comment`) VALUES 
(11937, 0, 0, 0, 54, 0, 100, 0, 0, 0, 0, 0, 49, 0, 0, 0, 0, 0, 0, 21, 30, 0, 0, 0, 0, 0, 0, 'Demon Portal Guardian - On Just summoned - Attack'),
(11937, 0, 1, 0, 6, 0, 100, 0, 0, 0, 0, 0, 41, 0, 0, 0, 0, 0, 0, 23, 0, 0, 0, 0, 0, 0, 0, 'Demon Portal Guardian - On death - Despawn summoner'),
(11937, 0, 2, 0, 21, 0, 100, 0, 0, 0, 0, 0, 106, 16, 0, 0, 0, 0, 0, 23, 0, 0, 0, 0, 0, 0, 0, 'Demon Portal Guardian - On homepos - remove flag'),
(11937, 0, 3, 0, 21, 0, 100, 0, 0, 0, 0, 0, 41, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 'Demon Portal Guardian - On homepos - Despawn');

jackpoz commented:
I will look into it, it might be it's the event itself (as it was bugged anyway)

jackpoz commented:
https://github.com/TrinityCore/TrinityCore/commit/507a56d62764f63113d35dfcca9273b8d0b16cc5 fixed, it was just the new hook not implemented in say
